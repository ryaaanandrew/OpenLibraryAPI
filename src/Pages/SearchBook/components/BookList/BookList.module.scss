.toolBar {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  margin: 2rem 0rem;
  padding: 1rem;
  border: 1px solid #bebebe;
  border-radius: 1rem;
  box-shadow: 0 5px 10px rgba(154, 160, 185, 0.05),
    0 15px 40px rgba(166, 173, 201, 0.2);

  &__item {
    margin-right: 2rem;
    cursor: pointer;

    & > h2 {
      margin: 0;
    }
  }
}

.bookItem {
  border: 1px solid #bebebe;
  padding: 2rem;
  border-radius: 1rem;
  box-shadow: 0 5px 10px rgba(154, 160, 185, 0.05),
    0 15px 40px rgba(166, 173, 201, 0.2);
  display: flex;
  flex-direction: row;
  justify-content: flex-start;

  &__details {
    margin-left: 2rem;
  }

  &__bookCoverPlaceholder {
    height: 224px;
    background-color: #dddddd;
    vertical-align: middle;

    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
  }

  &__bookCover {
    z-index: 10;
    height: 234px;
  }

  &__bookCover,
  &__bookCoverPlaceholder {
    border-radius: 0.5rem;
    width: 180px;
  }

  &:not(:last-of-type) {
    margin-bottom: 2rem;
  }
}

@media (max-width: 600px) {
  .toolBar {
    justify-content: space-evenly;
    text-align: center;

    &__item {
      margin: 0.5rem;
    }
  }

  .bookItem {
    flex-direction: column;
    align-items: center;

    &__details {
      text-align: center;
    }
  }
}
