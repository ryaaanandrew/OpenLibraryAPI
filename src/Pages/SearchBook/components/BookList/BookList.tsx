import { useEffect, useState } from "react";
import { sortBy, orderBy } from "lodash";
import styles from "../BookList/BookList.module.scss";
import moment from "moment";

const BookList: React.FC<{ data: any; isLoading: boolean }> = ({
  data,
  isLoading,
}) => {
  const [sortMethod, setSortMethod] = useState("");
  const [sortDirection, setSortDirection] = useState<"asc" | "desc">("asc");
  const {
    bookItem,
    bookItem__bookCoverPlaceholder,
    bookItem__bookCover,
    bookItem__details,
    toolBar,
    toolBar__item,
  } = styles;

  const sortedData = sortMethod
    ? orderBy(data, [sortMethod], [sortDirection])
    : data;

  const handleSort = (name: string) => {
    if (sortMethod === name) {
      if (sortDirection === "asc") setSortDirection("desc");
      if (sortDirection === "desc") setSortDirection("asc");
    }

    setSortMethod(name);
  };

  if (isLoading) return <h1>Loading...</h1>;

  return (
    <>
      {data.length === 0 ? (
        <h1>No books found</h1>
      ) : (
        <>
          <div className={toolBar}>
            <div className={toolBar__item} onClick={() => handleSort("title")}>
              <h2>Sort by title</h2>
            </div>
            <div
              className={toolBar__item}
              onClick={() => handleSort("publishDate")}
            >
              <h2>Sort by date</h2>
            </div>
            <div
              className={toolBar__item}
              onClick={() => handleSort("authorName")}
            >
              <h2>Sort by author</h2>
            </div>
          </div>
          {sortedData.map((item: any) => {
            return (
              <div key={item.key} className={bookItem}>
                <div className={bookItem__bookCoverPlaceholder}>
                  <img
                    src={`https://covers.openlibrary.org/b/isbn/${item.isbn?.[0]}-M.jpg`}
                    alt={`Book cover for ${item.title}`}
                    className={bookItem__bookCover}
                  />
                </div>

                <div className={bookItem__details}>
                  <h3>Title: {item.title}</h3>
                  <h3>Author: {item.author_name?.[0]}</h3>
                  <h3>
                    Publication Date:{" "}
                    {moment(item.publishDate).format("MMM, DD, YYYY")}
                  </h3>
                </div>
              </div>
            );
          })}
        </>
      )}
    </>
  );
};

export default BookList;
