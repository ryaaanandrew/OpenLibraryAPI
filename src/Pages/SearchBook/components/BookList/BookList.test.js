import { configureStore } from "@reduxjs/toolkit";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import rootReducer from "../../../../redux/store/store";
import BookList from "./BookList";

const store = configureStore({ reducer: rootReducer });

const mockBookList = [
  {
    key: "/works/OL25925967W",
    type: "work",
    seed: [
      "/books/OL35007332M",
      "/books/OL35027818M",
      "/books/OL35019737M",
      "/books/OL35005164M",
      "/books/OL34996504M",
      "/books/OL34994890M",
      "/books/OL35026134M",
      "/books/OL35028221M",
      "/books/OL34993054M",
      "/books/OL35011874M",
      "/books/OL35029684M",
      "/books/OL35016687M",
      "/books/OL35018410M",
      "/books/OL35013540M",
      "/books/OL35029890M",
      "/books/OL35021913M",
      "/books/OL34996393M",
      "/books/OL35004207M",
      "/works/OL25925967W",
      "/authors/OL9803489A",
    ],
    title: "Pokemon",
    title_suggest: "Pokemon",
    has_fulltext: false,
    edition_count: 18,
    edition_key: [
      "OL35007332M",
      "OL35027818M",
      "OL35019737M",
      "OL35005164M",
      "OL34996504M",
      "OL34994890M",
      "OL35026134M",
      "OL35028221M",
      "OL34993054M",
      "OL35011874M",
      "OL35029684M",
      "OL35016687M",
      "OL35018410M",
      "OL35013540M",
      "OL35029890M",
      "OL35021913M",
      "OL34996393M",
      "OL35004207M",
    ],
    publish_date: ["2021"],
    publish_year: [2021],
    first_publish_year: 2021,
    isbn: [
      "9798733458281",
      "9798733637907",
      "9798733162706",
      "9798733604015",
      "9798733192512",
      "9798733185330",
      "9798733471709",
      "9798733449562",
      "9798733459332",
      "9798733469935",
      "9798733480138",
      "9798733181271",
      "9798733453880",
      "9798733173634",
      "9798733151762",
      "9798733191911",
      "9798733625461",
      "9798733471853",
    ],
    last_modified_i: 1633825774,
    ebook_count_i: 0,
    publisher: ["Independently Published"],
    language: ["eng"],
    author_key: ["OL9803489A"],
    author_name: ["ARTpoppy fan"],
    publisher_facet: ["Independently Published"],
    _version_: 1715089692445114400,
    author_facet: ["OL9803489A ARTpoppy fan"],
    publishDate: "2021-01-01",
    authorName: "ARTpoppy fan",
  },
  {
    key: "/works/OL264247W",
    type: "work",
    seed: [
      "/books/OL12301716M",
      "/books/OL12301714M",
      "/books/OL12301713M",
      "/books/OL12301715M",
      "/books/OL12301719M",
      "/books/OL12301717M",
      "/books/OL12301718M",
      "/works/OL264247W",
      "/authors/OL2623771A",
    ],
    title: "Pokemon-e TCG",
    title_suggest: "Pokemon-e TCG",
    has_fulltext: false,
    edition_count: 7,
    edition_key: [
      "OL12301716M",
      "OL12301714M",
      "OL12301713M",
      "OL12301715M",
      "OL12301719M",
      "OL12301717M",
      "OL12301718M",
    ],
    publish_date: [
      "November 2003",
      "September 17, 2004",
      "September 17, 2003",
      "June 14, 2004",
    ],
    publish_year: [2004, 2003],
    first_publish_year: 2003,
    isbn: [
      "1930206917",
      "9781930206939",
      "193020681X",
      "9781930206847",
      "9781930206915",
      "1930206801",
      "1930206933",
      "1930206828",
      "9781930206809",
      "9781930206908",
      "9781930206816",
      "9781930206823",
      "1930206909",
      "1930206844",
    ],
    last_modified_i: 1304545050,
    ebook_count_i: 0,
    publisher: ["Nintendo Of America, Inc."],
    language: ["eng"],
    author_key: ["OL2623771A"],
    author_name: ["Various"],
    author_alternative_name: [
      "Varios",
      "Various Illustrators Various Authors",
      "VARIOS",
      "Varios Participantes",
      "varios",
      "varios authors",
      "VARIOS(132735)",
      "Author Autores Varios",
      "Various Non-Fiction Authors",
      "VARIOUS AUTHOR",
      "Autores varios",
      "Various.",
      "Various Celebrated Authors",
      "VARIOUS SPMG",
      "Varios autores",
      "Various, Authors",
      "Anthology: various authors",
      "Various AUTORES",
      "Varios Autores",
      "VARIOUS",
      "Various Chapter Authors",
      "Various Various",
      "various authors",
      "Various - DO NOT USE",
      "Varios Artistas",
      "VARIOS.",
      "Various Authors By",
      "Various Authors & Artists",
      "Various Aut.",
      "VARIOS(832172)",
      "Autores Varios",
      "Compendium of various authors",
      "Authors Various Authors",
      "A Compendium of Various Authors",
      "Various Authorities",
      "autores varios",
      "VARIOS AUTORES",
      "Compendium of Various Authors",
      "VARIOUS AUTHORS",
      "(various)",
      "Various Au",
      "by Various Authors",
      "Various Authors",
      "Vario",
      "Various authors",
      "Autores VARIOS",
      "various",
      "Various Aut",
      "Readers to Be Confirmed Various Authors",
      "Various Writers",
      "varios autores",
      "Various authors.",
      "various.",
    ],
    id_goodreads: ["2935214", "2821140", "2157956", "2608551"],
    publisher_facet: ["Nintendo Of America, Inc."],
    _version_: 1715090005701951500,
    author_facet: ["OL2623771A Various"],
    publishDate: "2003-11-01",
    authorName: "Various",
  },
  {
    key: "/works/OL25528094W",
    type: "work",
    seed: [
      "/books/OL34267629M",
      "/books/OL34247462M",
      "/books/OL34241324M",
      "/works/OL25528094W",
      "/authors/OL9646250A",
    ],
    title: "POKEMON Goo Malbuch",
    title_suggest: "POKEMON Goo Malbuch",
    has_fulltext: false,
    edition_count: 3,
    edition_key: ["OL34267629M", "OL34247462M", "OL34241324M"],
    publish_date: ["2021"],
    publish_year: [2021],
    first_publish_year: 2021,
    number_of_pages_median: 57,
    isbn: ["9798743934188", "9798743933389", "9798743934560"],
    last_modified_i: 1633109583,
    ebook_count_i: 0,
    publisher: ["Independently Published"],
    language: ["ger"],
    author_key: ["OL9646250A"],
    author_name: ["Pokemon POKEMON MALBUCH"],
    publisher_facet: ["Independently Published"],
    _version_: 1715089261140639700,
    author_facet: ["OL9646250A Pokemon POKEMON MALBUCH"],
    publishDate: "2021-01-01",
    authorName: "Pokemon POKEMON MALBUCH",
  },
];

test("SearchBook should match snapshot", () => {
  const view = render(
    <Provider store={store}>
      <BookList label={"Search"} data={mockBookList} />
    </Provider>
  );

  expect(view).toMatchSnapshot();
});
