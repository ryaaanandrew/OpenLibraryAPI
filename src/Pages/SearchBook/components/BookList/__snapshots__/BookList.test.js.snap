// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBook should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="toolBar"
      >
        <div
          class="toolBar__item"
        >
          <h2>
            Sort by title
          </h2>
        </div>
        <div
          class="toolBar__item"
        >
          <h2>
            Sort by date
          </h2>
        </div>
        <div
          class="toolBar__item"
        >
          <h2>
            Sort by author
          </h2>
        </div>
      </div>
      <div
        class="bookItem"
      >
        <div
          class="bookItem__bookCoverPlaceholder"
        >
          <h2
            class="bookItem__placeHolder"
          >
            Pokemon
          </h2>
          <img
            alt="Book cover for Pokemon"
            class="bookItem__bookCover"
            src="https://covers.openlibrary.org/b/isbn/9798733458281-M.jpg"
          />
        </div>
        <div
          class="bookItem__details"
        >
          <h3>
            Title: 
            Pokemon
          </h3>
          <h3>
            Author: 
            ARTpoppy fan
          </h3>
          <h3>
            Publication Date:
             
            Jan, 01, 2021
          </h3>
        </div>
      </div>
      <div
        class="bookItem"
      >
        <div
          class="bookItem__bookCoverPlaceholder"
        >
          <h2
            class="bookItem__placeHolder"
          >
            Pokemon-e TCG
          </h2>
          <img
            alt="Book cover for Pokemon-e TCG"
            class="bookItem__bookCover"
            src="https://covers.openlibrary.org/b/isbn/1930206917-M.jpg"
          />
        </div>
        <div
          class="bookItem__details"
        >
          <h3>
            Title: 
            Pokemon-e TCG
          </h3>
          <h3>
            Author: 
            Various
          </h3>
          <h3>
            Publication Date:
             
            Nov, 01, 2003
          </h3>
        </div>
      </div>
      <div
        class="bookItem"
      >
        <div
          class="bookItem__bookCoverPlaceholder"
        >
          <h2
            class="bookItem__placeHolder"
          >
            POKEMON Goo Malbuch
          </h2>
          <img
            alt="Book cover for POKEMON Goo Malbuch"
            class="bookItem__bookCover"
            src="https://covers.openlibrary.org/b/isbn/9798743934188-M.jpg"
          />
        </div>
        <div
          class="bookItem__details"
        >
          <h3>
            Title: 
            POKEMON Goo Malbuch
          </h3>
          <h3>
            Author: 
            Pokemon POKEMON MALBUCH
          </h3>
          <h3>
            Publication Date:
             
            Jan, 01, 2021
          </h3>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="toolBar"
    >
      <div
        class="toolBar__item"
      >
        <h2>
          Sort by title
        </h2>
      </div>
      <div
        class="toolBar__item"
      >
        <h2>
          Sort by date
        </h2>
      </div>
      <div
        class="toolBar__item"
      >
        <h2>
          Sort by author
        </h2>
      </div>
    </div>
    <div
      class="bookItem"
    >
      <div
        class="bookItem__bookCoverPlaceholder"
      >
        <h2
          class="bookItem__placeHolder"
        >
          Pokemon
        </h2>
        <img
          alt="Book cover for Pokemon"
          class="bookItem__bookCover"
          src="https://covers.openlibrary.org/b/isbn/9798733458281-M.jpg"
        />
      </div>
      <div
        class="bookItem__details"
      >
        <h3>
          Title: 
          Pokemon
        </h3>
        <h3>
          Author: 
          ARTpoppy fan
        </h3>
        <h3>
          Publication Date:
           
          Jan, 01, 2021
        </h3>
      </div>
    </div>
    <div
      class="bookItem"
    >
      <div
        class="bookItem__bookCoverPlaceholder"
      >
        <h2
          class="bookItem__placeHolder"
        >
          Pokemon-e TCG
        </h2>
        <img
          alt="Book cover for Pokemon-e TCG"
          class="bookItem__bookCover"
          src="https://covers.openlibrary.org/b/isbn/1930206917-M.jpg"
        />
      </div>
      <div
        class="bookItem__details"
      >
        <h3>
          Title: 
          Pokemon-e TCG
        </h3>
        <h3>
          Author: 
          Various
        </h3>
        <h3>
          Publication Date:
           
          Nov, 01, 2003
        </h3>
      </div>
    </div>
    <div
      class="bookItem"
    >
      <div
        class="bookItem__bookCoverPlaceholder"
      >
        <h2
          class="bookItem__placeHolder"
        >
          POKEMON Goo Malbuch
        </h2>
        <img
          alt="Book cover for POKEMON Goo Malbuch"
          class="bookItem__bookCover"
          src="https://covers.openlibrary.org/b/isbn/9798743934188-M.jpg"
        />
      </div>
      <div
        class="bookItem__details"
      >
        <h3>
          Title: 
          POKEMON Goo Malbuch
        </h3>
        <h3>
          Author: 
          Pokemon POKEMON MALBUCH
        </h3>
        <h3>
          Publication Date:
           
          Jan, 01, 2021
        </h3>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
